src/routing/effects/index.ts

```diff
@@ -0,0 +1,372 @@
+        routes,
+        location,
+        routesMap
+      ),
+      ({ data, urlMatcher }) => changeState({
+        data,
+        matcher: urlMatcher,
+        params,
+        historyChange: back ? 'back' : replace ? 'replace' : 'push',
+        location,
+      }),
+    )).value();
+
+/**
+ *  @param path    the current url in the browser
+ *  @param routes  all of the possible route states
+ *
+ *  @returns state
+ *
+ *  derieves a state from given path
+ *
+ */
+const navigate = (location: Location, routes, routesMap): any => {
+  const match = matchUrl(routes, location.path, location.query);
+
+  if (match.data.redirect && !match.data.data.redirectOnFailure) {
+    return Maybe.of(findState({ route: match.data.redirect, params: match.params, data: match.data.data },
+                              routes,
+                              location,
+                              routesMap,
+    )).map(pick(['data', 'params', 'pathname'])).value();
+  }
+
+  return match;
+
+};
+
+const checkLocation = ({ current, location, e, path }) => {
+  const newLocation = exists(e.state) ? e.state : matchUrl(current.routes, path, location.query);
+
+  return equals((current.location || location), newLocation) ? null : newLocation;
+};
+
+const handler = (e: PopStateEvent, current, path, location, routesMap) =>
+  Maybe.of(checkLocation({ current, location, e, path }))
+  .map(({ state, params, data: matchedState }) => {
+    if (matchedState && matchedState.state) {
+      return { data: matchedState, params, pathname: path };
+    }
+
+    const { data, urlMatcher } = getStateKey(JSON.parse(state), routesMap);
+
+    return { data, params, pathname: urlMatcher.format(merge(params, location.query)) };
+  }).value();
+
+export const splitQuery = pipe(
+  replaceStr(/^\?/, ''),
+  split('&'),
+  map(split('=')),
+  reduce((acc, [key, val]) => ({ ...acc, [key]: val }), {}),
+);
+
+const checkCurrent = (current, config) => current.location &&
+  !equals(current.location.state, config.current.state);
+
+// tslint:disable:max-func-body-length
+export default new Map<MessageConstructor, EffectHandler>([
+  [LoadToken, ({ key, result, error }, dispatch, execContext: ExecContext<any>) => {
+    const query = window.location.search !== '' ? splitQuery(window.location.search) : {};
+
+    if (query.authToken) {
+      const { authToken } = query;
+
+      try {
+        window.localStorage.setItem(key, safeStringify(safeParse(window.atob(authToken))));
+        const newQuery = new URLSearchParams(window.location.search);
+        newQuery.delete('authToken');
+        window.location.search = newQuery.toString();
+      } catch (e) {
+        // tslint:disable-next-line:no-console
+        console.warn(`Failure to store auth ${authToken}`);
+      }
+    }
+
+    dispatch(Message.construct(result, { key, value: safeParse(window.localStorage[key] || null) }));
+  }],
+
+  [Navigate, ({ ...props }, dispatch) => dispatch(new NavigationUpdated({
+    ...props,
```

Comment:
no need for spread

Reply:
This doesn't work without it

Reply:
That's weird. Did you try `new NavigationUpdated({ props })` or `new NavigationUpdated(props)`?